File: training-data/matlab/union_find.m
function [root, parent] = uf_find(parent, x)
    if parent(x) == x
        root = x;
    else
        [root, parent] = uf_find(parent, parent(x));
    end
end

function parent = uf_union(parent, x, y)
    [root_x, parent] = uf_find(parent, x);
    [root_y, parent] = uf_find(parent, y);
    if root_x ~= root_y
        parent(root_x) = root_y;
    end
end

parent = 1:5;
parent = uf_union(parent, 1, 2);
parent = uf_union(parent, 2, 3);
[r1, parent] = uf_find(parent, 1);
[r3, parent] = uf_find(parent, 3);
disp(r1 == r3)

File: training-data/prolog/union_find.pl
:- dynamic(parent/2).

init_uf(N) :-
    retractall(parent(_, _)),
    numlist(0, N, Nodes),
    forall(member(X, Nodes), assertz(parent(X, X))).

uf_find(X, X) :- parent(X, X), !.
uf_find(X, Root) :-
    parent(X, P),
    P \= X,
    uf_find(P, Root).

uf_union(X, Y) :-
    uf_find(X, RootX),
    uf_find(Y, RootY),
    (RootX \= RootY -> retract(parent(RootX, RootX)), assertz(parent(RootX, RootY)) ; true).

:- init_uf(5), uf_union(0, 1), uf_union(1, 2),
   uf_find(0, R1), uf_find(2, R2), uf_find(3, R3),
   (R1 =:= R2 -> writeln(true) ; writeln(false)),
   (R1 =:= R3 -> writeln(true) ; writeln(false)).
